omniscidb/ResultSetRegistry/ResultSetRegistry.cpp

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ResultSetRegistry.h"
+
+#include "DataMgr/DataMgr.h"
+
+#include <iomanip>
+
+namespace hdk {
+
+namespace {
+
+class ResultSetDataToken : public Data_Namespace::AbstractDataToken {
+ public:
+  ResultSetDataToken(ResultSetPtr rs, const int8_t* buf, size_t size)
+      : rs_(std::move(rs)), buf_(buf), size_(size) {}
+  ~ResultSetDataToken() override {}
+
+  const int8_t* getMemoryPtr() const override { return buf_; }
+  size_t getSize() const override { return size_; }
+
+ private:
+  ResultSetPtr rs_;
+  const int8_t* buf_;
+  size_t size_;
+};
+
+TableFragmentsInfo getEmptyTableMetadata(int table_id) {
+  TableFragmentsInfo res;
+  res.setPhysicalNumTuples(0);
+
+  // Executor requires dummy empty fragment for empty tables
+  FragmentInfo& empty_frag = res.fragments.emplace_back();
+  empty_frag.fragmentId = 0;
+  empty_frag.shadowNumTuples = 0;
+  empty_frag.setPhysicalNumTuples(0);
+  empty_frag.deviceIds.push_back(0);  // Data_Namespace::DISK_LEVEL
+  empty_frag.deviceIds.push_back(0);  // Data_Namespace::CPU_LEVEL
+  empty_frag.deviceIds.push_back(0);  // Data_Namespace::GPU_LEVEL
+  empty_frag.physicalTableId = table_id;
+  res.fragments.push_back(empty_frag);
+
+  return res;
+}
+
+}  // namespace
+
+ResultSetRegistry::ResultSetRegistry(ConfigPtr config,
+                                     const std::string& schema_name,
+                                     int db_id)
+    : SimpleSchemaProvider(hdk::ir::Context::defaultCtx(),
+                           getSchemaId(db_id),
+                           schema_name)
+    , db_id_(db_id)
+    , schema_id_(getSchemaId(db_id))
+    , config_(config) {}
+
+std::shared_ptr<ResultSetRegistry> ResultSetRegistry::init(
+    Data_Namespace::DataMgr* data_mgr,
+    ConfigPtr config) {
+  static mapd_shared_mutex res_registry_init_mutex;
+  mapd_shared_lock<mapd_shared_mutex> check_init_lock(res_registry_init_mutex);
+
+  auto* ps_mgr = data_mgr->getPersistentStorageMgr();
+  if (!ps_mgr->hasDataProvider(hdk::ResultSetRegistry::SCHEMA_ID)) {
+    check_init_lock.unlock();
+    mapd_unique_lock<mapd_shared_mutex> init_lock(res_registry_init_mutex);
+    if (!ps_mgr->hasDataProvider(hdk::ResultSetRegistry::SCHEMA_ID)) {
+      ps_mgr->registerDataProvider(hdk::ResultSetRegistry::SCHEMA_ID,
+                                   std::make_shared<hdk::ResultSetRegistry>(config));
+    }
+  }
+
+  auto provider = ps_mgr->getDataProvider(hdk::ResultSetRegistry::SCHEMA_ID);
```

Comment:
This happens to be under the lock as well. Is it intended?

Reply:
It didn't bother me much to have this getter under the shared lock. But you are right it's not necessary. I reduced the scope of the lock.